### ISSUE TITLE
CompCorrAuto

### ISSUE THREAD

**Issue by Zhisen** [stnava/ANTs#192]
Hi, we are trying to extract time series from the csf and white matter from a functional mri file so that later we can use csf and white matter as two covariates in our regression program. 

1). We ran Atropos to segment the image into three tissue types with the following command: Atropos -d 3 -x anat_noskull_masked.nii -c [3,0] -m [1,1x1x1] -i KMeans[3] -o anat_seg.nii -a anat_noskull.nii. 

2). We wanted to use the segmented file to extract time series for csf and white matter. We ran the following command but it kept giving us a "segmentation fault" error message: ImageMath 4 output.nii ThreeTissueConfounds 4D_TimeSeries.nii anat_seg.nii 1 3.  Do we have the usage wrong? Also, is it possible to run ThreeTissueConfounds with an image that has more than 3 image labels?

3). Once we get the output from step 2, how do we use CompCorrAuto to proceed? What is the usage of CompCorrAuto?

Thanks a lot!

Zhisen

**Comment by stnava** [comment]
we moved most of our fmri image processing tools to [ANTsR](http://stnava.github.io/ANTsR/) ... you might check the code to see how these functions work.   here are some links:

CompCorrAuto : https://github.com/stnava/ANTs/blob/master/Examples/ImageMath.cxx#L3963

ThreeTissueConfounds : https://github.com/stnava/ANTs/blob/master/Examples/ImageMath.cxx#L4254 

@jeffduda  and @bkandel are the guys who know the fmri processing in ANTsR best ... e.g. how to most quickly get ROI/Label summary measures from time series matrices.  there are both ASL and resting state fMRI vignettes in the works (+fairly mature).

**Comment by bkandel** [comment]
Agreed--this sort of thing is exactly what ANTsR is made for.  Extracting a
time-series from a 4D image and a 3D mask can be accomplished by
timeseries2matrix, and you can write that to a CSV or whatever for later
processing.

2015-06-10 18:33 GMT-04:00 stnava <notifications@github.com>:

**Comment by stnava** [comment]
so you might do something like this in antsr?

```
labels=antsImageRead("labelimage.nii.gz")
csfmask=maskImage( labels, 1 )
csfsummary=rowMeans(  timeseries2matrix( boldImage, csfmask ) )
```

yes?

**Comment by bkandel** [comment]
Yes, exactly.

2015-06-10 18:41 GMT-04:00 stnava <notifications@github.com>:

**Comment by Zhisen** [comment]
Hi, 
   Thank you for the replies. Does this mean that CompCorrAuto can only be ran on ANTsR now, or it still can be ran on regular ANTS? If it is the later, I think I have one very focused question, that is, what are inputs and in what order should I put them after the CompCorrAuto below the general command of ImageMath. 
   The reason I have to ask this is that we have the documentation about ImageMath, which explained how to use ImageMath itself and also how to use each of the Operators, except that it just did not explain how to use CompCorrAuto. For example, for the operator of TimeSeriesSubset, it describes what the output would be and also says Usage: TimeSeriesSubset 4D_TimeSeries..nii.gz. However, there is no such "Usage" for CompCorrAuto. We are at the stage where we just need that one command to work before proceeding to the next step. 

   If all has to be done on R now, then we will move to R for it. 
   Thanks a lot!

Best,
Marina

**Comment by stnava** [comment]
we recommend using R for this.  otherwise you can read the code and add the documentation yourself.

here is the code:

CompCorrAuto : https://github.com/stnava/ANTs/blob/master/Examples/ImageMath.cxx#L3963

if you make changes and want us to incorporate them, please make a pull request.

https://help.github.com/articles/using-pull-requests/

**Comment by Zhisen** [comment]
Hi, 
  Now we are using ANTsR and keep getting this error message. Can you tell us what is wrong here. Thanks a lot! 

 Error message: 
    Error in apply(mat[, maskvec == labs[1]], FUN = mean, MARGIN = 1) : 
  dim(X) must have a positive length
The command we ran: 
 compcor(fmri, ncompcor=4, variance_extreme=0.975, mask=anat, 
+         randomSamples=1, returnv=F, returnhighvarmat=F, returnhighvarmatinds=F,
+         highvarmatinds=NA)


Best,
Marina

**Comment by dorianps** [comment]
If i remember correctly the margin comes first,  then the function.  Check
the help ?apply

**Comment by Zhisen** [comment]
yes, apply(X, MARGIN, FUN, ...)
how do I fix it then? Thanks!

**Comment by stnava** [comment]
here is the basic approach

```
img=antsImageRead("fmri.nii.gz")
mask=getMask(img)
mat=timeseries2matrix( img, mask )
compcorrdf<-compcor( mat )
```

**Comment by bkandel** [comment]
Yes, the order is not the problem here.  Can you give us your images?  The
example does work for me:
 mat <- matrix( rnorm(50000) ,ncol=500)
     compcorrdf<-compcor( mat )
It is possible, as Brian suggested, that you are putting in the raw image
and not the time series matrix into compcor.

2015-06-16 14:06 GMT-04:00 stnava <notifications@github.com>:

**Comment by dorianps** [comment]
apply(x, 1, mean) should work,  or rowMeans()

**Comment by Zhisen** [comment]
Yes  I am putting in the raw fmri image, see attached. Oh, it is a big file, it can't attach. 

I just ran this, as you suggested:
img<-antsImageRead("fm_run1_al.nii") 
mask<-getMask("anat_3.5.nii")
mat=timeseries2matrix(img, mask)
compcorrdf<-compcor(mat)
And here is the error message: Error in compcor(mat) : object 'mat' not found

**Comment by Zhisen** [comment]
The fmri file is 267.50mb, too big to be sent.

**Comment by cookpa** [comment]
It's not a problem with ordering of arguments.

R parameters are only positional if you are not using param=value. So

  apply(mat[, maskvec == labs[1]], FUN = mean, MARGIN = 1)

is OK, as is

  apply(mat[, maskvec == labs[1]], MARGIN = 1,  FUN = mean )

The error message suggests something is wrong with the "mat" object

  > Error in apply(mat[, maskvec == labs[1]], FUN = mean, MARGIN = 1) :
  > dim(X) must have a positive length

**Comment by cookpa** [comment]
Try 

  antsImageRead("fm_run1_al.nii", 4)

**Comment by Zhisen** [comment]
just try importing fmri image with dimension of 4, the same error message as object 'mat' not bound

**Comment by stnava** [comment]
this line won't work

```
mask<-getMask("anat_3.5.nii")
```

try running this:

```
img=antsImageRead("fm_run1_al.nii")
mask=getMask(img)
mat=timeseries2matrix( img, mask )
compcorrdf<-compcor( mat )
```

**Comment by bkandel** [comment]
Looks like the problem is that you haven't read the anatomical image in.

2015-06-16 14:13 GMT-04:00 Philip Cook <notifications@github.com>:

> Try
>
> antsImageRead("fm_run1_al.nii", 4)
>
>

**Comment by Zhisen** [comment]
It seems to me it reads it in. I tried to class it, and it identifies it. 

> class(mask)
[1] "antsImage"
attr(,"package")
[1] "ANTsR"

**Comment by Zhisen** [comment]
ok, more detailed error message here: 
img<-antsImageRead("fm_run1_al.nii",4) 
> mask<-getMask("anat_3.5.nii")
Error in getMask("anat_3.5.nii") : 
  'lowthresh' and 'highthresh' must be numeric scalars
In addition: Warning message:
In mean.default(img) : argument is not numeric or logical: returning NA
> mat<-timeseries2matrix(img, mask)
Error in apply(mat[, maskvec == labs[1]], FUN = mean, MARGIN = 1) : 
  dim(X) must have a positive length
> compcorrdf<-compcor(mat)
Error in compcor(mat) : object 'mat' not found

**Comment by stnava** [comment]
this line won't work

```
mask<-getMask("anat_3.5.nii")
```

try running this:

```
img=antsImageRead("fm_run1_al.nii")
mask=getMask(img)
mat=timeseries2matrix( img, mask )
compcorrdf<-compcor( mat )
```

**Comment by Zhisen** [comment]
Is it because of the name of the file "anat_3.5.nii" ?

**Comment by cookpa** [comment]
Ah OK - you need to give getMask an antsImage, I think:

  # Dimension should be auto detected
  maskImg<-antsImageRead("anat_3.5.nii")

  mask<-getMask(maskImg)

If you don't specify thresholds, it will mask from the mean of maskImg to its maximum

**Comment by stnava** [comment]
just run this exactly:

```
img=antsImageRead("fm_run1_al.nii")
mask=getMask(img)
mat=timeseries2matrix( img, mask )
compcorrdf<-compcor( mat )
```     
then try the version with the anatomical image .... but i am guessing your anatomical image is not aligned with your bold image so that probably won't work.  basically, the mask needs to be in the same space as the fmri.  usually, the anatomical image is not in the same space unless you've done some registration and resampling first.

**Comment by Zhisen** [comment]
The anatomical image is aligned with the functional image. the functional image ends with "al" because it was aligned to that anatomical image. so I don't think that is the problem. They are also at the same resolution.

**Comment by Zhisen** [comment]
I just tried this:
img<-antsImageRead("fm_run1_al.nii") 
> maskImg<-antsImageRead("anat_3.5.nii")
> mask<-getMask(maskImg)
> mat<-timeseries2matrix(img, mask)
> compcorrdf<-compcor(mat)

no error message this time, but I also see no new files created or old files modified. nothing happened.

**Comment by dorianps** [comment]
Not sure what you expect to be saved, but ANTsR is not the same as ANTs. Objects and images are stored in variables. Unless you save the variable it won't save itself.

**Comment by Zhisen** [comment]
Oh, now I actually see the matrix that is named compcorrdf. So it seems it works now. The question now is how do I use this file. My understanding was that by running compcor, i will be returned with a functional file (image) that have the csf and white matter confounds filtered out. Is this file now the filtered file? Does it mean i need to convert it into an image file? Yeah, I do realize this difference between ANTs and ANTsR, the file format difference.

**Comment by Zhisen** [comment]
Ok, more understanding. It seems this is not the approach where compcor extracts the high temporal standard deviation voxels based on functional data to filter out csf and white matter. It seems what this is doing now, with timeseries2matrix, it extacts the timeseries of the mask. For this, I should have used the csf mask for extracting time series for csf, and the white matter mask for extracting time series for white matter. Does that sound correct?

**Comment by stnava** [comment]
compcor is usually run on the whole brain.  but you can use it in several different ways, depending on your interests.  e.g. you can run it on the background image to model the background.  you might also find that `preprocessfMRI` is valuable.

```
set.seed(123)
     n=16
     nvox <- n*n*n*12
     dims <- c(n,n,n,12)
     boldImage <- makeImage(dims, rnorm(nvox) + 500) %>% iMath("PadImage", 2)
     # for real data: boldImage <- antsImageRead(getANTsRData('pcasl'))
     cleanfMRI <- preprocessfMRI(boldImage)
```

**Comment by Zhisen** [comment]
Ok, I see. Thanks a lot!

And thank you so much to everyone that gave me good advice. The prompt responses were like live chat. That was great!

**Comment by Zhisen** [comment]
Hi all, 

  I have to ask a follow-up question. I mentioned this earlier but probably did not make it clear enough about what I want. So what I am hoping to get here is to regress out the time series of csf and white matter from a functional mri data. The best for my subsequent analysis is to get an output file that is the functional mri data but with the csf and wm time series already regressed out, that is, a cleaned functional data file. Right now, the output matrix is the time series of the mask, be it the csf or the wm. What code can I use to get the functional output that already regressed out the mask's matrix? Thanks a lot!

**Comment by stnava** [comment]
`preprocessfMRI`


brian

**Comment by Zhisen** [comment]
*Thank you for pointing to preprocessfMRI. It would give the output I want.
So, I ran it, and it seems to all worked except that writing the image out
into my working directory does not seem to work. This is more a general
question about how to write an image file out than a question about
preprocessing itself. Still, I am showing the steps I took below as an
example,  and pointing out where my question is. Help is very appreciated!*

*#preprocess, it worked.*
img<-antsImageRead("fm_run1_al.nii")
maskImg<-antsImageRead("anat_3.5.nii")
mask<-getMask(maskImg)
cleanfMRI<-preprocessfMRI(img, maskImage=mask, maskingMeanRatioThreshold=NA,
                          initialNuisanceVariables= NA,
                          numberOfCompCorComponents=6,doMotionCorrection=T,

useMotionCorrectedImage=F,motionCorrectionAccuracyLevel=1,
                          meanBoldFixedImageForMotionCorrection=NA,

frequencyLowThreshold=NA,frequencyHighThreshold=NA,

spatialSmoothingType="none",spatialSmoothingParameters=0,
                          residualizeMatrix=T)

*#convert matrxi into ANTsImage, it worked*
cleanimg<-makeImage(mask,cleanfMRI,spacing=c(3.5, 3.5, 3.5),

origin=c(NA),direction=c(NA),components=1,pixeltype='float')

*# I ran the command below, but no new file shows up in my working
directory. And I was not given any error message when running this command.
What do you think went wrong with this command? Thanks a lot! *
antsImageWrite(cleanimg,tempfile(pattern="fm_run1_al_corrected",fileext=".nii.gz"))



*Thanks a lot!*
*Marina*

**Comment by stnava** [comment]
to write an image, do:

```
antsImageWrite( img,  filename)
```

that's all.

you should not be doing this:

```
cleanimg<-makeImage(mask,cleanfMRI,spacing=c(3.5, 3.5, 3.5),

origin=c(NA),direction=c(NA),components=1,pixeltype='float')
```

if you follow the example that is in the help, you would do:

```

     set.seed(123)
     n=16
     nvox <- n*n*n*12
     dims <- c(n,n,n,12)
     boldImage <- makeImage(dims, rnorm(nvox) + 500) %>% iMath("PadImage", 2)
     # for real data: boldImage <- antsImageRead(getANTsRData('pcasl'))
     cleanfMRI <- preprocessfMRI(boldImage)
  
# produces 
names(cleanfMRI)
# [1] "cleanBoldImage"    "maskImage"         "DVARS"            
# [4] "DVARSpostCleaning" "FD"                "globalSignal"     
# [7] "nuisanceVariables"
cleanfMRI$cleanBoldImage
# antsImage
#  Pixel Type          : float 
# Components Per Pixel: 1 
#  Dimensions          : 20x20x20x16 
#  Voxel Spacing       : 1x1x1x1 
#  Origin              : -2 -2 -2 -2 
#  Direction           : 1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 1 
antsImageWrite(  cleanfMRI$cleanBoldImage , "temp.nii.gz" )
```

**Comment by Zhisen** [comment]
putting  $cleanBoldImage after cleanfMRI seems to do that trick. Without
it, R returns the message that cleanfMRI is not ANTsImage, and that was why
I was doing makeImage. I did not see in any help file that says to use
$cleanBoldImage. Can you point to me where to find it? Thanks!

**Comment by Zhisen** [comment]
Now I can visualize the preprocessed boldimage in afni, it turns out the
almost all the signals are lost after preprocessing. Any number or argument
settings that I should change in preprocessfMRI to improve this?

**Comment by Zhisen** [comment]
maybe i was confused by what $cleanBoldImage was doing and ended up writing
out the wrong file.

**Comment by stnava** [comment]
you need to set up the parameters you want for your experiment.

`cleanBoldImage` is the cleaned image - i.e. remove nuisance variables.

so it wont look like a "standard" bold image anymore.


brian

**Comment by Zhisen** [comment]
I actually see no signal at all when I graph the time series of the
processed image. If only the nuisance variables were regressed out, then
there should be remaining bold signals. I am attaching the image (not the
graph). the left is before preprocessing, where the yellowish orange part
shows the bold signals + nuisance activity. The right is the image after
preprocessing, where everything is green, meaning almost no value there. I
would assume most of the originally-yellowish voxels should at least
maintain some values.

**Comment by stnava** [comment]
can't help with this unless you share data.

**Comment by Zhisen** [comment]
I would like to share the data with you and get your help. I am new with ANTsR.
Right now nifty is not one of the files supported by this discussion board for uploading. Can you tell me how to send you the data? Thanks!

**Comment by stnava** [comment]
wetransfer.com


brian

**Comment by Zhisen** [comment]
Hi Brain,

   I transferred both the bold image file and the mask. Thank you so much
for your help!

Best,
Zhisen

**Comment by stnava** [comment]
great - but don't see a link or email .... would be helpful if you posted the link.  in the mean time, @jeffduda is working on a resting state vignette over at [https://github.com/stnava/ANTsR](https://github.com/stnava/ANTsR) and here is some code that shows more or less what you might do with the preprocessing function.   this builds a bit on what @jeffduda supplies in the basic vignette.

```
library(ANTsR)
  img = antsImageRead(getANTsRData("rsbold"))
  mask = antsImageRead(getANTsRData("rsboldmask"))
  seg = antsImageRead(getANTsRData("rsboldseg"))
  # Identify first steady-state timepoint
  tr = spacing[4]
  steady = ceiling(10.0 / tr)
  # Eliminate non steady-state timepoints, restore image info
  mat=timeseries2matrix( img, mask )
  mat=mat[steady:nrow(mat),]
  img=matrix2timeseries(img,mask,mat)
  meanbold <- apply.antsImage(img, c(1,2,3), mean)
pp=preprocessfMRI( img, maskImage=mask,
  numberOfCompCorComponents=8,
   doMotionCorrection=TRUE,
   residualizeMatrix=FALSE,
   spatialSmoothingParameters=3.0  ) # very little smoothing b/c it's in spacing
mat=timeseries2matrix( pp$cleanBoldImage  , mask )
rmat=residuals(lm( mat ~ 1 + pp$FD + pp$nuisanceVariables )) # get the clean image
eanat=sparseDecom( scale(rmat,scale=FALSE), mask, smooth=0.0,  #decompose into parts
  nvecs=50, sparseness=0.01, its=3, cthresh=250, mycoption=0)
eseg=eigSeg( mask, eanat$eig )  # identify the parts
rimg=matrix2timeseries( img, mask, rmat )
plot( getAverageOfTimeSeries( img ) , eseg, axis=3 )  
jeanat=joinEigenanatomy( datamatrix=rmat, mask=mask, listEanatImages=eanat$eigenanatomyimages, graphdensity=c(5:10/100))  # join the parts by network-based community detection 
jeseg=eigSeg( mask, jeanat$fusedlist )
plot( getAverageOfTimeSeries( img ) , jeseg, axis=3 )  # look at the communities in brain space
pheatmap::pheatmap( cor( eanat$projections ) )
```

we use [http://igraph.org/r/](http://igraph.org/r/) for any graph/network calculations.

**Comment by Zhisen** [comment]
Ok it seems the email notification did not reach you. Here is the download link:http://we.tl/y7kMRNH72r

fm_run1_al is the bold signal, and anat.3.5is the anatomy to serve as mask


Thanks a lot!

Best,
Zhisen

**Comment by stnava** [comment]
ok - the most crucial thing here is that the "mask" image that you sent is not a mask.  nor is it even the brain alone ... it is the whole head anatomical image.  obviously, it's more common that you would isolate the cerebrum and segment it as well before processing your bold data.   although i personally am happy to process bold data without any t1 at all.   here is an example of how to do that:

```
library(ANTsR)
mask = antsImageRead( "anat_3.5.nii.gz" )  # useless - not a mask
img = antsImageRead( "fm_run1_al.nii.gz" ) # get mask from here
meanbold = getAverageOfTimeSeries( img )  # this is the mean bold
mask = getMask( meanbold )                 # this is the mask
#  seg = antsImageRead(getANTsRData("rsboldseg")) # if you had a segmentation
# Identify first steady-state timepoint
spacing = antsGetSpacing(img)
tr = spacing[4]
steady = ceiling( 10.0 / tr )
mat = timeseries2matrix( img, mask )
mat = mat[steady:nrow(mat),]
img = matrix2timeseries(img,mask,mat)
# check out a time slice 
image(  as.array( img )[,30,15,] ) # not too much motion
img = smoothImage( img, spacing, sigmaInPhysicalCoordinates = TRUE, FWHM = FALSE)
pp = preprocessfMRI( img, maskImage=mask,
   numberOfCompCorComponents=6, # reduce to as low as 2 or 4 depending on data
   doMotionCorrection=0,   # should be higher if motion is severe 
   residualizeMatrix=FALSE,
   spatialSmoothingParameters=0.0 )
mat = timeseries2matrix( pp$cleanBoldImage  , mask )
print( colnames( pp$nuisanceVariables ) )
rmat = residuals(lm( mat ~ 0 + pp$nuisanceVariables )) # remove nuisance from raw bold
rimg = matrix2timeseries( img, mask, rmat )
plot( getAverageOfTimeSeries( rimg ) ,  axis=2, dorot=1 ) # wont look standard anymore
# here i think about 1/2 of the brain will have meaningful signal so let us parcellate that
# into small chunks covering about 5 percent each
eanat = sparseDecom( rmat, mask, smooth=0.,
  nvecs=10, sparseness=0.05, its=3, cthresh=250, mycoption=0)
eseg = eigSeg( mask, eanat$eig )
plot( meanbold, eseg,  axis=2, dorot=1 )
# make a graph
p1 = abs( imageListToMatrix( eanat$eig, mask) )
p1 = p1/rowSums( data.matrix(p1) )
p1 = rmat %*% t(p1)
pheatmap::pheatmap( antsrimpute( cor( p1 ) ) ) # if this looks good then ....
grph = makeGraph( cor(p1) )
```

obviously this data-driven graph may be difficult to interpret --- that is often the cost of not using strong priors.  however, what we gain from this is that the signal explaining high variance is largely in cortex - as we expect after nuisance regression.

a 2nd note is that the residualized bold image wont look like "standard bold" anymore because you're thrown a bunch of signal away.  there is no one size fits all solution for denoising bold as it is rather objective specific:  i.e. resting state vs decoding vs task-based activation studies may all require different processing choices which should be informed by your knowledge of the problem.

**Comment by Zhisen** [comment]
Hi Brian,

 Thank you so much for the very detailed description. I will get into it
and may come up with questions for more clarification.

 Thanks a whole lot again!

Best,
Zhisen

**Comment by Zhisen** [comment]
Hi Brian,

    The commands ran quite well. I have one question regarding smoothing.
We don't want any more smoothing in this process, so I took out this :

img = smoothImage( img, spacing, sigmaInPhysicalCoordinates = TRUE,
 FWHM = FALSE)

  And then in preprocessfMRI, I leave the spatialSmoothingParameters
  as it is, i.e., 0.0. However, it turns out that the output file have
 1.1 mm more smoothing than the input file. Do you have any idea how
  this happens and how to prevent this from happening?

I am going to copy what I ran below. By the way, the mask created from
the functional file ended up cutting out the lateral temporal cortex (due

to lower signal there). We want to keep the lateral temporal cortex. Thus
I used an anatomy file that has only the brain (no skull, no other
structures) to create the mask.

img=antsImageRead("fm_run1_al.nii")
maskImg<-antsImageRead("anat_noskull_3.5.nii")
mask<-getMask(maskImg)
spacing=antsGetSpacing(img)
tr=spacing[4]
steady=ceiling(10.0/tr)
mat=timeseries2matrix(img,mask)
mat=mat[steady:nrow(mat),]
img=matrix2timeseries(img,mask,mat)
pp=preprocessfMRI(img,maskImage=mask,numberOfCompCorComponents=6,doMotionCorrection=0,
                  residualizeMatrix=F,spatialSmoothingParameters=0.0)
mat=timeseries2matrix(pp$cleanBoldImage,mask)
rmat=residuals(lm(mat~0 + pp$nuisanceVariables))
rimg=matrix2timeseries(img,mask,rmat)
antsImageWrite(rimg, "fm_run1_al_clean.nii.gz")

It would be great if there is a way to not incur smoothing when implementing

the commands above. Thanks a lot!

Best,

Zhisen

**Comment by stnava** [comment]
That all sounds fine. I don't think there is any additional smoothing. you
may be seeing the effect of interpolation.  Feel free to look through the
code to see if you can find any other smoothing but I doubt there is any...

**Comment by Zhisen** [comment]
Hi Brian,

   I looked into the code, and it seems the only smoothing is the
spatialSmoothingType, which by default is 'none'. We used the default, and
also set smoothing parameters to 0. Thus, it does not seem it is smoothing
that was done in the process.

  With 'interpolation', what exactly does it mean in ANTsR? Is it some
autocorrelation that is added in this process? If so, is there a way to
prevent it? Any autocorrelation would be counted as smoothing, which we do
not want. I know that the output and the input files have the same
resolution, so it does not seem there was resampling for that purpose.

  Thanks a lot!

Best,
Zhisen
